File: tools/tasks/changelog.py
import re, shutil, sys
from tools import common
from datetime import date
from pathlib import Path

def _load_codeowners_usernames(path: Path = Path(".github") / "CODEOWNERS") -> set[str]:
    """
    Return a set of GitHub usernames (no leading @) found in CODEOWNERS.
    If file doesn't exist, returns empty set.
    """
    try:
        text = path.read_text(encoding="utf-8")
    except FileNotFoundError:
        return set()
    # @user OR @org/team → capture the user/team token(s). We only use bare usernames.
    names = set(re.findall(r'@([A-Za-z0-9][A-Za-z0-9\-\/]+)', text))
    # Keep only bare usernames (drop org/team like org/team)
    return {n for n in names if "/" not in n}

def _gh_commit_author_login(owner_repo: str, sha: str) -> str | None:
    """
    Uses `gh` to resolve the GitHub author login for a commit.
    Returns login (str) or None if unknown.
    """
    code, out, err = common.run([
        "gh", "api", f"repos/{owner_repo}/commits/{sha}",
        "--jq", ".author.login // empty"
    ])
    if code != 0:
        return None
    return out.strip() or None

def _version_key(v: str) -> tuple[int, ...]:
    """
    Very light numeric comparator for versions like 0.1.10 vs 0.1.5.
    Extracts all integers and compares lexicographically.
    """
    nums = [int(x) for x in common.regex.CHANGELOG_VERSION_TOKEN.findall(v)]
    return tuple(nums) if nums else (0,)

def _assert_not_lower_than_highest(changelog_path: Path, version: str):
    """
    Fail early if `version` is lower than the highest version heading in the CHANGELOG.
    Looks for headings like: ## [0.1.2] - YYYY-MM-DD
    """
    if not changelog_path.is_file():
        return  # let other checks handle missing file
    text = changelog_path.read_text(encoding="utf-8")
    found = [m.group(1) for m in re.finditer(common.regex.CHANGELOG_VERSION_HEADER, text)]
    # filter out 'Unreleased' if present accidentally
    found = [v for v in found if v.lower() != "unreleased"]
    if not found:
        return
    highest = max(found, key=_version_key)
    if _version_key(version) < _version_key(highest):
        common.fail(f"Error: provided version {version} is lower than highest in CHANGELOG ({highest}).")

def _update_changelog_file(changelog_path: Path, version: str, notes_md: str, owner_repo: str | None):
    if not changelog_path.is_file():
        common.fail(f'Error: CHANGELOG not found at "{changelog_path}".')

    text = changelog_path.read_text(encoding="utf-8")
    today = date.today().strftime("%Y-%m-%d")

    # 1) Upsert the version section
    new_block = f"## [{version}] - {today}\n\n{notes_md.strip()}\n\n"
    m = common.regex.CHANGELOG_HEADER_FOR(version).search(text)

    if m:
        start = m.start()
        next_header = re.search(common.regex.CHANGELOG_NEXT_HEADER, text[m.end():])
        end = m.end() + (next_header.start() if next_header else len(text) - m.end())
        updated = text[:start] + new_block + text[end:]
    # else:
    #     first_header = re.search(common.regex.CHANGELOG_NEXT_HEADER, text)
    #     if first_header:
    #         insert_at = first_header.start()
    #         updated = text[:insert_at] + new_block + text[insert_at:]
    #     else:
    #         updated = text.rstrip() + "\n\n" + new_block
    else:
        # If there's an Unreleased section, insert the new version *after* it and its content,
        # i.e., right before the first actual version header following Unreleased.
        unreleased_hdr = re.search(common.regex.CHANGELOG_UNRELEASED_CHANGES_HEADER, text)
        if unreleased_hdr:
            next_after_unreleased = re.search(common.regex.CHANGELOG_NEXT_HEADER, text[unreleased_hdr.end():])
            if next_after_unreleased:
                insert_at = unreleased_hdr.end() + next_after_unreleased.start()
            else:
                insert_at = len(text)
            updated = text[:insert_at] + new_block + text[insert_at:]
        else:
            # No Unreleased section: insert before the very first version header (top of history)
            first_header = re.search(common.regex.CHANGELOG_NEXT_HEADER, text)
            if first_header:
                insert_at = first_header.start()
                updated = text[:insert_at] + new_block + text[insert_at:]
            else:
                updated = text.rstrip() + "\n\n" + new_block

    # 2) Update the [Unreleased] compare link
    if not owner_repo:
        owner_repo = common.parse_owner_repo_from_origin()
    if not owner_repo:
        common.fail("Error: Unable to determine owner/repo from 'origin' to update [Unreleased] link.")

    unreleased_line = f"[Unreleased]: https://github.com/{owner_repo}/compare/v{version}...HEAD"
    m_unreleased = common.regex.CHANGELOG_UNRELEASED_REF.search(updated)
    if m_unreleased:
        # Replace exactly one line
        start, end = m_unreleased.span()
        updated = updated[:start] + unreleased_line + updated[end:]
        insert_after = updated.find("\n", start)
        insert_pos = (insert_after + 1) if insert_after != -1 else len(updated)
    else:
        # If not present, append and set insertion point after it
        updated = updated.rstrip() + "\n" + unreleased_line + "\n"
        insert_pos = len(updated)

    # 3) Ensure the version reference line exists, and place it *below Unreleased*
    ref_exists = bool(common.regex.REF_LINE_FOR(version).search(updated))
    if not ref_exists:
        vers_ref = f"[{version}]: https://github.com/{owner_repo}/releases/tag/v{version}"
        updated = updated[:insert_pos] + vers_ref + "\n" + updated[insert_pos:]

    changelog_path.write_text(updated, encoding="utf-8")

def changelog(release):
    version = common.normalize_version(release)
    if shutil.which("git") is None:
        common.fail("Error: git is not on PATH.")

    changelog_path = Path("CHANGELOG.md")
    _assert_not_lower_than_highest(changelog_path, version)

    # Ensure remote refs fresh
    common.run(["git", "fetch", "--prune", "origin"])

    owner_repo = common.parse_owner_repo_from_origin() or ""
    codeowners = _load_codeowners_usernames()

    # Collect SHA|subject in develop not in origin/stable
    code, out, err = common.run([
        "git", "log",
        "--pretty=%H|%s",
        "--reverse",
        "origin/stable..develop",
        "--no-merges"
    ])
    if code != 0:
        common.fail(f"Error: git log failed. Details: {err or out or 'unknown'}")

    raw_lines = [l for l in out.splitlines() if l.strip()]
    if not raw_lines:
        print("(no new commits on develop relative to origin/stable)", file=sys.stderr)
        sys.exit(0)

    lines: list[str] = []
    for row in raw_lines:
        try:
            sha, subj = row.split("|", 1)
        except ValueError:
            sha, subj = "", row

        login = _gh_commit_author_login(owner_repo, sha) if sha and owner_repo else None
        if login and login not in codeowners:
            subj = f"{subj} ({login})"
        lines.append(subj)

    notes_md = "\n".join(f"- {s}" for s in lines)
    _update_changelog_file(changelog_path, version, notes_md, owner_repo=owner_repo)

    print(f"CHANGELOG.md updated for [{version}] with {len(lines)} commit(s).")

File: tools/tasks/changelog.js
import fs from 'node:fs';
import path from 'node:path';
import * as common from '../common.js';

console.log(`Loaded: ${import.meta.url}`);

/**
 * Return a set of GitHub usernames (no leading @) found in CODEOWNERS.
 * If the file doesn't exist, returns an empty set.
 * @param {string} [codeownersPath]
 * @returns {Set<string>}
 */
function loadCodeownersUsernames(codeownersPath = path.join('.github', 'CODEOWNERS')) {
    let text;
    try {
        text = fs.readFileSync(codeownersPath, 'utf-8');
    } catch {
        return new Set();
    }
    const names = new Set();
    for (const m of text.matchAll(/@([A-Za-z0-9][A-Za-z0-9\-/]+)/g)) names.add(m[1]);
    return new Set([...names].filter(n => !n.includes('/')));
}

/**
 * Uses `gh` to resolve the GitHub author login for a commit.
 * @param {string} ownerRepo
 * @param {string} sha
 * @returns {string|null}
 */
function ghCommitAuthorLogin(ownerRepo, sha) {
    const {code, stdout} = common.run([
        'gh', 'api', `repos/${ownerRepo}/commits/${sha}`,
        '--jq', '.author.login // empty',
    ]);
    if (code !== 0) return null;
    return stdout.trim() || null;
}

/**
 * Very light numeric comparator for versions like 0.1.10 vs 0.1.5.
 * Extracts all integers and compares lexicographically.
 * @param {string} v
 * @returns {number[]}
 */
function versionKey(v) {
    const nums = [...v.matchAll(common.regex.CHANGELOG_VERSION_TOKEN)].map(m => parseInt(m[0], 10));
    return nums.length ? nums : [0];
}

function compareVersionKeys(a, b) {
    const len = Math.max(a.length, b.length);
    for (let i = 0; i < len; i++) {
        const av = a[i] ?? 0, bv = b[i] ?? 0;
        if (av !== bv) return av - bv;
    }
    return 0;
}

/**
 * Fail early if `version` is lower than the highest version heading in the CHANGELOG.
 * Looks for headings like: ## [0.1.2] - YYYY-MM-DD
 * @param {string} changelogPath
 * @param {string} version
 */
function assertNotLowerThanHighest(changelogPath, version) {
    if (!fs.existsSync(changelogPath)) return; // let other checks handle missing file
    const text = fs.readFileSync(changelogPath, 'utf-8');
    const found = [...text.matchAll(common.regex.CHANGELOG_VERSION_HEADER)]
        .map(m => m[1])
        .filter(v => v.toLowerCase() !== 'unreleased');
    if (!found.length) return;
    let highest = found[0];
    for (const v of found) if (compareVersionKeys(versionKey(v), versionKey(highest)) > 0) highest = v;
    if (compareVersionKeys(versionKey(version), versionKey(highest)) < 0) {
        common.fail(`Error: provided version ${version} is lower than highest in CHANGELOG (${highest}).`);
    }
}

/**
 * @param {string} changelogPath
 * @param {string} version
 * @param {string} notesMd
 * @param {string|null} ownerRepo
 */
function updateChangelogFile(changelogPath, version, notesMd, ownerRepo) {
    if (!fs.existsSync(changelogPath)) {
        common.fail(`Error: CHANGELOG not found at "${changelogPath}".`);
    }

    let text = fs.readFileSync(changelogPath, 'utf-8');
    const now = new Date();
    const today = `${now.getFullYear()}-${String(now.getMonth() + 1).padStart(2, '0')}-${String(now.getDate()).padStart(2, '0')}`;

    // 1) Upsert the version section
    const newBlock = `## [${version}] - ${today}\n\n${notesMd.trim()}\n\n`;
    const headerRe = common.regex.CHANGELOG_HEADER_FOR(version);
    const m = headerRe.exec(text);

    let updated;
    if (m) {
        const start = m.index;
        const end0 = m.index + m[0].length;
        const rest = text.slice(end0);
        const nextHeaderRe = new RegExp(common.regex.CHANGELOG_NEXT_HEADER.source, common.regex.CHANGELOG_NEXT_HEADER.flags);
        const nextHeader = nextHeaderRe.exec(rest);
        const end = end0 + (nextHeader ? nextHeader.index : rest.length);
        updated = text.slice(0, start) + newBlock + text.slice(end);
    } else {
        // If there's an Unreleased section, insert the new version *after* it and
        // its content, i.e. right before the first actual version header following it.
        const unreleasedRe = new RegExp(common.regex.CHANGELOG_UNRELEASED_CHANGES_HEADER.source, common.regex.CHANGELOG_UNRELEASED_CHANGES_HEADER.flags);
        const unreleasedHdr = unreleasedRe.exec(text);
        if (unreleasedHdr) {
            const afterUnreleased = text.slice(unreleasedHdr.index + unreleasedHdr[0].length);
            const nextHeaderRe = new RegExp(common.regex.CHANGELOG_NEXT_HEADER.source, common.regex.CHANGELOG_NEXT_HEADER.flags);
            const nextAfterUnreleased = nextHeaderRe.exec(afterUnreleased);
            const insertAt = nextAfterUnreleased
                ? (unreleasedHdr.index + unreleasedHdr[0].length) + nextAfterUnreleased.index
                : text.length;
            updated = text.slice(0, insertAt) + newBlock + text.slice(insertAt);
        } else {
            // No Unreleased section: insert before the very first version header (top of history)
            const firstHeaderRe = new RegExp(common.regex.CHANGELOG_NEXT_HEADER.source, common.regex.CHANGELOG_NEXT_HEADER.flags);
            const firstHeader = firstHeaderRe.exec(text);
            if (firstHeader) {
                updated = text.slice(0, firstHeader.index) + newBlock + text.slice(firstHeader.index);
            } else {
                updated = text.replace(/\s+$/, '') + '\n\n' + newBlock;
            }
        }
    }

    // 2) Update the [Unreleased] compare link
    if (!ownerRepo) ownerRepo = common.parseOwnerRepoFromOrigin();
    if (!ownerRepo) {
        common.fail("Error: Unable to determine owner/repo from 'origin' to update [Unreleased] link.");
    }

    const unreleasedLine = `[Unreleased]: https://github.com/${ownerRepo}/compare/v${version}...HEAD`;
    const unreleasedRefRe = new RegExp(common.regex.CHANGELOG_UNRELEASED_REF.source, common.regex.CHANGELOG_UNRELEASED_REF.flags);
    const mUnreleased = unreleasedRefRe.exec(updated);
    let insertPos;
    if (mUnreleased) {
        const start = mUnreleased.index;
        const end = start + mUnreleased[0].length;
        updated = updated.slice(0, start) + unreleasedLine + updated.slice(end);
        const insertAfter = updated.indexOf('\n', start);
        insertPos = insertAfter !== -1 ? insertAfter + 1 : updated.length;
    } else {
        updated = updated.replace(/\s+$/, '') + '\n' + unreleasedLine + '\n';
        insertPos = updated.length;
    }

    // 3) Ensure the version reference line exists, and place it *below Unreleased*
    const refExistsRe = common.regex.REF_LINE_FOR(version);
    const refExists = refExistsRe.test(updated);
    if (!refExists) {
        const versRef = `[${version}]: https://github.com/${ownerRepo}/releases/tag/v${version}`;
        updated = updated.slice(0, insertPos) + versRef + '\n' + updated.slice(insertPos);
    }

    fs.writeFileSync(changelogPath, updated, 'utf-8');
}

/**
 * @param {string} release - release tag or version string
 */
export async function changelog(release) {
    const version = common.normalizeVersion(release);
    const {code: gitCode} = common.run(['git', '--version']);
    if (gitCode !== 0) common.fail('Error: git is not on PATH.');

    const changelogPath = path.join('CHANGELOG.md');
    assertNotLowerThanHighest(changelogPath, version);

    // Ensure remote refs fresh
    common.run(['git', 'fetch', '--prune', 'origin']);

    const ownerRepo = common.parseOwnerRepoFromOrigin() || '';
    const codeowners = loadCodeownersUsernames();

    // Collect SHA|subject in develop not in origin/stable
    const {code, stdout, stderr} = common.run([
        'git', 'log',
        '--pretty=%H|%s',
        '--reverse',
        'origin/stable..develop',
        '--no-merges',
    ]);
    if (code !== 0) common.fail(`Error: git log failed. Details: ${stderr || stdout || 'unknown'}`);

    const rawLines = stdout.split('\n').filter(l => l.trim());
    if (!rawLines.length) {
        console.error('(no new commits on develop relative to origin/stable)');
        process.exit(0);
    }

    const lines = [];
    for (const row of rawLines) {
        const sepIdx = row.indexOf('|');
        const sha = sepIdx === -1 ? '' : row.slice(0, sepIdx);
        let subj = sepIdx === -1 ? row : row.slice(sepIdx + 1);

        const login = (sha && ownerRepo) ? ghCommitAuthorLogin(ownerRepo, sha) : null;
        if (login && !codeowners.has(login)) subj = `${subj} (${login})`;
        lines.push(subj);
    }

    const notesMd = lines.map(s => `- ${s}`).join('\n');
    updateChangelogFile(changelogPath, version, notesMd, ownerRepo || null);

    console.log(`CHANGELOG.md updated for [${version}] with ${lines.length} commit(s).`);
}
